### PR TITLE
Work for 1143

### PR DESCRIPTION
**Pull request by sbrossie**

### CODE REVIEW

**Comment by sbrossie** [comment]
FYI: Integration tests passed locally:

```
Finished in 1123.122646 seconds.

185 tests, 10056 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

```

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/overdue/TestOverdueIntegration.java [hunk 1]
```diff
@@ -1142,7 +1142,7 @@ public void testOverdueClearWithCredit() throws Exception {
 
         final BigDecimal accountBalance = invoiceUserApi.getAccountBalance(account.getId(), callContext);
 
-        busHandler.pushExpectedEvents(NextEvent.INVOICE, NextEvent.BLOCK);
+        busHandler.pushExpectedEvents(NextEvent.INVOICE, NextEvent.INVOICE_ADJUSTMENT, NextEvent.INVOICE_ADJUSTMENT, NextEvent.BLOCK);
```
Comment (pierre) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/overdue/TestOverdueIntegration.java]: Why two `INVOICE_ADJUSTMENT` events (might be worth adding a comment for the future reader...)?

File: invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java [hunk 2]
```diff
@@ -425,23 +428,26 @@ public void createInvoice(final InvoiceModelDao invoice,
...
-                        // Should we maybe enforce callers (e.g. InvoiceApiHelper) to properly populate these invoices?
-                        cbaDao.doCBAComplexityFromTransaction(adjustedInvoiceId, invoicesTags, entitySqlDaoWrapperFactory, context);
-                    }
-
-                    if (adjustedCommittedInvoiceIds.contains(adjustedInvoiceId)) {
-                        // Notify the bus since the balance of the invoice changed (only if the invoice is COMMITTED)
-                        notifyBusOfInvoiceAdjustment(entitySqlDaoWrapperFactory, adjustedInvoiceId, accountId, context.getUserToken(), context);
+                // CBA COMPLEXITY...
+                //
+                // Optimized path where we don't need t refresh invoices
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java]: Typo (`to`)

File: invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java [hunk 3]
```diff
@@ -767,8 +774,11 @@ public boolean apply(final InvoicePaymentModelDao input) {
                 // Notify the bus since the balance of the invoice changed
                 final UUID accountId = transactional.getAccountIdFromInvoicePaymentId(chargeBack.getId().toString(), context);
 
-                cbaDao.doCBAComplexityFromTransaction(payment.getInvoiceId(), invoicesTags, entitySqlDaoWrapperFactory, context);
-
+                final Set<UUID> modifiedInvoiceIds = new HashSet<>();
+                modifiedInvoiceIds.addAll(cbaDao.doCBAComplexityFromTransaction(ImmutableSet.of(payment.getInvoiceId()), invoicesTags, entitySqlDaoWrapperFactory, context));
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java]: Nit: the pattern

1. create set to track ids
2. run CBA
3. notify bus

repeats itself a few times - might be possible to abstract a function?

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/overdue/TestOverdueIntegration.java (hunk 1 above)
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/overdue/TestOverdueIntegration.java]: See cb5ae6d104cf

File: invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java (hunk 3 above)
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java]: See cb5ae6d104cf

**Pull request merged**